Userland/Libraries/LibVideo/VP9/TreeParser.cpp

```diff
@@ -346,13 +346,94 @@ ErrorOr<ReferenceMode> TreeParser::parse_comp_mode(BitStream& bit_stream, Probab
+        } else if (left_intra.value()) {
+            if (above_single.value()) {
+                context = 1 + 2 * (above_ref_frame_0.value() != comp_var_ref[1]);
+            } else {
+                context = 1 + 2 * (above_ref_frame_biased.value() != comp_var_ref[1]);
+            }
+        } else if (above_intra.value()) {
+            if (left_single.value()) {
+                context = 1 + 2 * (left_ref_frame_0.value() != comp_var_ref[1]);
+            } else {
+                context = 1 + 2 * (left_ref_frame_biased != comp_var_ref[1]);
+            }
+        } else {
+            auto var_ref_above = above_single.value() ? above_ref_frame_0 : above_ref_frame_biased;
+            auto var_ref_left = left_single.value() ? left_ref_frame_0 : left_ref_frame_biased;
+            if (var_ref_above == var_ref_left && comp_var_ref[1] == var_ref_above) {
+                context = 0;
+            } else if (left_single.value() && above_single.value()) {
+                if ((var_ref_above == comp_fixed_ref && var_ref_left == comp_var_ref[0])
+                    || (var_ref_left == comp_fixed_ref && var_ref_above == comp_var_ref[0])) {
+                    context = 4;
+                } else if (var_ref_above == var_ref_left) {
+                    context = 3;
+                } else {
+                    context = 1;
+                }
+            } else if (left_single.value() || above_single.value()) {
+                auto vrfc = left_single.value() ? var_ref_above : var_ref_left;
+                auto rfs = above_single.value() ? var_ref_above : var_ref_left;
+                if (vrfc == comp_var_ref[1] && rfs != comp_var_ref[1]) {
+                    context = 1;
+                } else if (rfs == comp_var_ref[1] && vrfc != comp_var_ref[1]) {
+                    context = 2;
+                } else {
+                    context = 4;
+                }
+            } else if (var_ref_above == var_ref_left) {
+                context = 4;
+            } else {
+                context = 2;
+            }
+        }
+    } else if (above_intra.has_value()) {
+        if (above_intra.value()) {
+            context = 2;
+        } else {
+            if (above_single.value()) {
+                context = 3 * static_cast<u8>(above_ref_frame_0.value() != comp_var_ref[1]);
+            } else {
+                context = 4 * static_cast<u8>(above_ref_frame_biased.value() != comp_var_ref[1]);
+            }
+        }
+    } else if (left_intra.has_value()) {
+        if (left_intra.value()) {
+            context = 2;
+        } else {
+            if (left_single.value()) {
+                context = 3 * static_cast<u8>(left_ref_frame_0.value() != comp_var_ref[1]);
+            } else {
+                context = 4 * static_cast<u8>(left_ref_frame_biased != comp_var_ref[1]);
+            }
+        }
+    } else {
+        context = 2;
+    }
+
+    u8 probability = probability_table.comp_ref_prob()[context];
```

Comment:
non-blocking comment, but the meaning of "context" is inscrutable in this function

Reply:
Definitely true, but given that the contexts for these probabilities are most likely fairly arbitrary, I'm not sure that I could come up with a better name. Once the above and left contexts have a struct, though, it'll at least be possible to give those parameters better names without the function declaration getting out of hand. That'll be one of the things I'll look into when I get to that.